# Human vote prediction -- the human comparison for RQ1

**Aim.** Establish what the same observable persuasion profile predicts about a
*human* vote, so the LLM surrogate results have a human reference point. This
is a comparison analysis, deliberately smaller than the LLM pipeline: three
feature sets, one estimator, one coefficient figure.

The notebook has two independent parts.

**Part A -- Lai et al. (2023) replication.** Reproduces the original binary
pairwise vote-prediction experiment on the released train/val/test split, as a
sanity check that our reconstruction of the corpus is sound. It is not
optimised and nothing downstream depends on it.

**Part B -- thesis human-vote protocol.** The same binary outcome, but with the
191-game cleaned corpus, the 13 persuasion features validated for the LLM
surrogate, and the dependence-aware evaluation protocol used throughout the
thesis: repeated grouped nested CV over the 36 player compositions, with a
composition-clustered bootstrap for every interval.

Fixed design decisions:

- Outcome: $y_{ABg} = 1$ when human voter $A$ votes for candidate $B$, for
  every ordered pair $A \neq B$ in game $g$.
- **No ground-truth roles anywhere in Part B.** The LLM observer cannot see
  start roles, end roles, or who the werewolf was, so the human comparison
  cannot either. Lai's role-inclusive model is not reproduced.
- Features: the 13 RQ1 persuasion features, as per-100-turn rates, built by
  reusing `utils_choice.rq1_features` rather than reimplementing annotation
  parsing.
- Estimator: $L_2$-penalised logistic regression, `class_weight="balanced"`.
  Only $C$ is tuned, by inner grouped CV on ROC-AUC. No class-weight search,
  no threshold tuning.
- Primary metric: held-out ROC-AUC from pair-level averaged out-of-fold
  probabilities. Secondary: F1 at the fixed 0.5 threshold.

Associations, not effects: these are regularised coefficients from an
observational corpus, and the bootstrap intervals are intervals, not
hypothesis tests.

## Setup

In [1]:
import sys, time
from pathlib import Path

import numpy as np
import pandas as pd


def find_repo_root(start=None, repo_name="masters_thesis_sdg"):
    current = (start or Path.cwd()).resolve()
    while True:
        if current.name == repo_name:
            return current
        if current.parent == current:
            raise FileNotFoundError(repo_name)
        current = current.parent


REPO_ROOT = find_repo_root()
if str(REPO_ROOT / "src") not in sys.path:
    sys.path.insert(0, str(REPO_ROOT / "src"))

from utils_choice.io import load_vote_tables
from utils_choice.rq1_features import (build_feature_table, ALL_FEATURES_13,
                                       STRATEGY_FEATURES, ENRICHED_FEATURES)
from utils_choice.rq1_model import add_rate_columns
from utils_choice.rq1_cv import (build_outer_folds, outer_splits, inner_fold_map,
                                 N_OUTER_FOLDS, N_INNER_FOLDS, OUTER_SEEDS)
from utils_logreg.lai2023_loading import load_annotation_splits, load_outcome_index

import utils_choice.human_lai as lai
import utils_choice.human_pairs as hp
import utils_choice.human_model as hm

MODEL_STAGE, PROMPT_DIR = "base", "prompt_v4"
ANALYSIS_ROOT = REPO_ROOT / "analysis"
TABLES_REL = Path(MODEL_STAGE) / "voting" / PROMPT_DIR / "vote_stability" / "tables"
ANNOT_ROOT = REPO_ROOT / "data" / "raw" / "lai2023"
ACC_ROOT = (REPO_ROOT / "data" / "processed" / "lai2023"
            / "accusation_transcripts" / "acc_targets")
IC_ROOT = (REPO_ROOT / "data" / "processed" / "lai2023"
           / "identity_claim_transcripts" / "ic_targets")

#: the validated LLM RQ1 outputs -- read only, never written by this notebook
RQ1_TABLES = (ANALYSIS_ROOT / "cross_model" / MODEL_STAGE / "voting" / PROMPT_DIR
              / "predictive_v2" / "tables")

#: new, versioned output directory; the previous human analysis under
#: pairwise_logreg/ is left untouched for comparison
OUT_ROOT = ANALYSIS_ROOT / "human_outcomes" / "pairwise_logreg_v2"
TAB, FIG, DIA = OUT_ROOT / "tables", OUT_ROOT / "figures", OUT_ROOT / "diagnostics"
for d in (TAB, FIG, DIA):
    d.mkdir(parents=True, exist_ok=True)

N_BOOT = hm.N_BOOT
pd.set_option("display.width", 200)
print("repo:", REPO_ROOT)
print("writing to:", OUT_ROOT.relative_to(REPO_ROOT))

repo: C:\Users\annab\Documents\GitHub\masters_thesis_sdg
writing to: analysis\human_outcomes\pairwise_logreg_v2


## 2. Data and validation

Corpus, roster and features all come from the same loaders the LLM RQ1
pipeline uses, so the two analyses cannot silently diverge.

In [2]:
votes, games, roster = load_vote_tables(ANALYSIS_ROOT, TABLES_REL)
annot_splits = load_annotation_splits(ANNOT_ROOT)
outcome_index = load_outcome_index(ANNOT_ROOT)

counts = hp.corpus_counts(roster)
assert counts == {"games": 191, "unique_players": 60, "unique_compositions": 36}, counts
print("corpus OK:", counts)

# the roster the LLM was shown and Lai's playerNames must be the same people
from utils_choice.features import ckey
from utils_logreg.lai2023_loading import get_session_key, get_game_id
n_checked = 0
for dataset, by_split in annot_splits.items():
    for split, glist in by_split.items():
        for game in glist:
            key = ckey(hp.DATASET_SOURCE[dataset], get_session_key(game, dataset),
                       get_game_id(game))
            assert key in roster, f"annotation game missing from the roster: {key}"
            rec = outcome_index[dataset][get_session_key(game, dataset)][get_game_id(game)]
            assert set(rec["playerNames"]) == set(roster[key]["players"]), key
            n_checked += 1
assert n_checked == 191
print(f"roster agreement OK: all {n_checked} annotated games match the LLM roster exactly")

corpus OK: {'games': 191, 'unique_players': 60, 'unique_compositions': 36}
roster agreement OK: all 191 annotated games match the LLM roster exactly


In [3]:
feat_df, roster, nmaps, diagnostics = build_feature_table(
    ANALYSIS_ROOT, TABLES_REL, ANNOT_ROOT, ACC_ROOT, IC_ROOT)
feat_df, zero_denominator = add_rate_columns(feat_df)
print(f"player-game feature rows: {len(feat_df)} over {feat_df.key.nunique()} games")

# the 13 features must be numerically identical to the validated RQ1 table
support_check = hp.assert_matches_rq1_support(feat_df, RQ1_TABLES / "feature_support_final.csv")
print("feature totals reproduce the RQ1 feature table exactly for all 13 features")

conservation = hp.conservation_report(feat_df)
print(f"early + late == overall for all {len(conservation)} counted quantities")

unresolved = hp.unresolved_name_summary(diagnostics)
print("\nunresolved speaker names (identical to RQ1):")
print(unresolved.to_string(index=False))
assert set(unresolved["raw_name"]) == {"Jordan", "Kayla", "Matt"}, \
    "unresolved-name set differs from the RQ1 pipeline"
print(f"\nzero-denominator rate rows: {len(zero_denominator)} "
      "(numerator asserted zero in every case, so rate 0 is exact)")

# no ground-truth role or outcome may reach the thesis feature matrices
_forbidden = ("role", "werewolf_ground", "won", "correct", "end_role", "start_role")
_leaks = [c for c in ALL_FEATURES_13 if any(t in c for t in _forbidden)
          and not c.startswith(("claims_", "werewolf_accusations", "deception_accusations"))]
assert not _leaks, f"ground-truth-looking column in the feature list: {_leaks}"
print("no ground-truth role, outcome or correctness field is in the 13-feature list")

player-game feature rows: 864 over 191 games
feature totals reproduce the RQ1 feature table exactly for all 13 features
early + late == overall for all 14 counted quantities

unresolved speaker names (identical to RQ1):
                source raw_name  n
      original_speaker   Jordan 95
     accusation_target   Jordan 22
identity_claim_speaker   Jordan 15
    accusation_accuser   Jordan 13
      original_speaker     Matt  9
      original_speaker    Kayla  3
identity_claim_speaker    Kayla  1

zero-denominator rate rows: 275 (numerator asserted zero in every case, so rate 0 is exact)
no ground-truth role, outcome or correctness field is in the 13-feature list


## 3. Lai et al. (2023) replication

Lai et al. report 2741 / 427 / 827 train/val/test rows. Building all $n^2$
ordered pairs from the released split files gives 2741 / **445** / **836**.
The whole discrepancy is one rule, recovered from the authors' released code
(`SALT-NLP/PersuationGames`, `baselines/read_data.py`): players whose
`startRoles` entry is `Moderator` are skipped on **both** the voter and the
candidate side. There are exactly three such players in the corpus, all in
5-player Ego4D games -- two in val, one in test -- and each costs
$25 - 16 = 9$ rows:

$$\text{val } 445 - 18 = 427, \qquad \text{test } 836 - 9 = 827.$$

Two further details also come from the released code, not the paper:

- **Self-pairs are kept.** The loops do not skip $i = j$, so a game
  contributes $n^2$ rows, not $n(n-1)$. 34 `votingOutcome` entries are in fact
  self-votes and are scored as positives here.
- **Abstainers stay in the design matrix.** `votingOutcome` records a
  non-voter as the string `"NA"` / `"N/A"` (41 cases) and twice as an
  out-of-range integer; the released code maps those to an impossible index,
  so the voter contributes $n$ all-negative rows instead of being dropped.

So the corpus is *not* the source of the discrepancy: our cleaned 191-game
ONUW set is exactly Lai's (199 games minus the 8 Avalon games), with identical
rosters, and both replicate their counts to the row.

**Hyperparameters and AUC.** `C=1.4, class_weight={0:1, 1:4.2}` is hardcoded
in the authors' `main_deduction.py` above a commented-out search, so it is a
released-code value, not one stated in the paper -- calling it "paper exact"
was wrong. Their AUC is `roc_auc_score(y_true, y_score=model.predict(X))`,
which passes **hard 0/1 labels** where a score is expected; the reported 54.7%
is therefore a balanced accuracy, not a ROC-AUC. Both are reported below:
`auc_hard` to match their number, `auc_prob` as the standard quantity used
everywhere else in this thesis.

Only the strategies-only model is reproduced. The role-inclusive variant is
excluded because ground-truth start roles are precisely what the LLM observer
does not have.

In [4]:
lai_rows = lai.build_lai_split_rows(annot_splits, outcome_index)
lai_counts = lai.split_row_counts(lai_rows)
print(lai_counts.to_string(index=False))
assert lai_counts["matches_paper"].all(), \
    "row counts no longer reproduce Lai et al.'s 2741 / 427 / 827"

lai_result = lai.run_lai_model(lai_rows, C=1.4, class_weight={0: 1.0, 1: 4.2})
lai_replication = lai.replication_table({"Lai Strategy": lai_result}, lai_rows)
print()
print(lai_replication.to_string(index=False))

split  rows  positives  prevalence  paper_rows  matches_paper
train  2741        575    0.209777        2741           True
  val   427         87    0.203747         427           True
 test   827        152    0.183797         827           True



                    model  paper_f1  our_f1  paper_auc  our_auc_prob  our_auc_hard_as_released  n_features  n_train  n_val  n_test  test_prevalence
             Lai Strategy     0.322  0.3273      0.546        0.5545                    0.5541        14.0     2741    427     827           0.1838
Random prediction (paper)     0.286     NaN      0.500           NaN                       NaN         NaN     2741    427     827              NaN


## 4. Thesis human-vote protocol

Same binary outcome, different -- and stricter -- construction:

| | Replication (Part A) | Thesis protocol (Part B) |
|---|---|---|
| pairs per game | $n^2$, self-pairs kept | $n(n-1)$, `voter != candidate` |
| abstaining voters | kept, all-negative rows | excluded, and logged |
| features | 7 Lai strategy proportions per side | 13 RQ1 persuasion rates per side |
| voter role | available | never |
| evaluation | the released fixed split | repeated grouped nested CV |

A voter enters only when their `votingOutcome` entry is a valid index of a
*different* roster player, which guarantees exactly one positive among their
$n-1$ rows. Every roster player stays available as a **candidate**, including
moderators and abstainers, because that is exactly the alternative set the LLM
was shown for the same game.

In [5]:
pairs, excluded_voters = hp.build_pair_rows(annot_splits, outcome_index, roster)
structure = hp.assert_pairwise_structure(pairs, roster)
print("pairwise structure OK:", structure)
print(f"\npositive prevalence: {structure['prevalence']:.4f}")
print("\nexcluded voters (never candidates -- they stay on every ballot):")
print(excluded_voters.groupby("reason").size().to_string())
print(f"\ngames contributing at least one voter: {pairs.key.nunique()} of {len(roster)}")

pairwise structure OK: {'pair_rows': 2882, 'positives': 784, 'prevalence': 0.27203331020124916, 'voters': 784, 'games': 184}

positive prevalence: 0.2720

excluded voters (never candidates -- they stay on every ballot):
reason
moderator                   3
no_vote_recorded           41
self_vote                  34
vote_index_out_of_range     2

games contributing at least one voter: 184 of 191


### Feature representation

Rates, not counts, matching the LLM rate representation exactly: speaker-side
features are normalised by that player's annotated turns, received-accusation
features by the turns produced by *all other* rostered players, both expressed
per 100 turns. Zero own turns implies a zero numerator and hence rate 0, which
`add_rate_columns` asserts rather than assumes; a received-accusation rate
never depends on the candidate's own turn count.

Three feature sets, and only three. Their columns come from
`rq1_model.feature_columns`, the same function the LLM models use, so they
cannot drift apart. `No Strategy` appears only in Part A -- it is part of
Lai's 7-dimensional simplex representation, and has no place in a rate model
where the six strategies are already independent per-turn rates.

In [6]:
for name, (representation, block) in hp.FEATURE_SETS.items():
    cols = hp.feature_set_columns(name)
    print(f"{name:20s} {representation:14s} {block:9s} "
          f"{len(cols):2d} per side -> {2 * len(cols):2d} predictors")

assert len(hp.feature_set_columns("Strategy")) == 6
assert len(hp.feature_set_columns("Strategy Temporal")) == 12
assert len(hp.feature_set_columns("Combined")) == 13
print("\nCombined columns:")
for c in hp.feature_set_columns("Combined"):
    print("  ", c)

Strategy             rate_overall   strategy   6 per side -> 12 predictors
Strategy Temporal    rate_temporal  strategy  12 per side -> 24 predictors
Combined             rate_overall   combined  13 per side -> 26 predictors

Combined columns:
   accusation_rate
   defense_rate
   interrogation_rate
   evidence_rate
   identity_declaration_rate
   call_for_action_rate
   werewolf_accusations_made_rate
   deception_accusations_made_rate
   werewolf_accusations_received_rate
   deception_accusations_received_rate
   claims_werewolf_rate
   claims_tanner_rate
   claims_night_action_role_rate


### Grouped nested cross-validation

The fold assignments are taken verbatim from `rq1_cv.build_outer_folds`, the
same deterministic composition-to-fold map the LLM analysis uses, so human and
LLM models are scored on the same held-out game structure: 5 outer folds, 5
repeated partitions (seeds 0-4), 3-fold grouped inner CV, grouping by the 36
unique player compositions.

This controls repeated *compositions*. It does **not** give participant-level
independence: 22 of the 60 players recur across different compositions, so an
individual can still appear on both sides of a split via a different game.
That limitation is inherent to the corpus.

In [7]:
fold_df = build_outer_folds(roster, n_folds=N_OUTER_FOLDS, seeds=OUTER_SEEDS)
_, comp_of_key = outer_splits(fold_df, repeat=0)
pairs["composition_id"] = [comp_of_key[k] for k in pairs["key"]]
assert pairs["composition_id"].nunique() == 36

# the fold map must be the one the LLM analysis used
rq1_folds = pd.read_csv(RQ1_TABLES / "group_cv_assignments.csv")
ours = fold_df[["repeat", "fold", "source", "session", "game", "composition_id"]]
theirs = rq1_folds[["repeat", "fold", "source", "session", "game", "composition_id"]]
sort_cols = ["repeat", "source", "session", "game"]
pd.testing.assert_frame_equal(ours.sort_values(sort_cols).reset_index(drop=True),
                              theirs.sort_values(sort_cols).reset_index(drop=True),
                              check_dtype=False)
print("fold assignments are identical to the LLM RQ1 pipeline's "
      f"({len(fold_df)} rows, {fold_df.repeat.nunique()} repeats x {fold_df.key.nunique()} games)")

# explicit leakage assertions, per repeat and per fold
for rep in sorted(fold_df["repeat"].unique()):
    test_keys_by_fold, comp_r = outer_splits(fold_df, rep)
    seen = set()
    for f, test_keys in test_keys_by_fold.items():
        te = pairs[pairs["key"].isin(test_keys)]
        tr = pairs[~pairs["key"].isin(test_keys)]
        assert not (set(te["composition_id"]) & set(tr["composition_id"])), \
            f"repeat {rep} fold {f}: a composition spans train and test"
        assert not (set(te["key"]) & set(tr["key"])), \
            f"repeat {rep} fold {f}: a game spans train and test"
        # inner CV is generated from the outer-training games only
        inner = inner_fold_map(set(tr["key"].unique()), comp_r,
                               n_folds=N_INNER_FOLDS, seed=rep * 100 + f)
        assert not (set(inner) & test_keys), \
            f"repeat {rep} fold {f}: an outer-test game reached the inner CV"
        seen |= set(te["key"])
    assert seen == set(pairs["key"]), f"repeat {rep}: a game was never held out"
print("CV assertions OK: no composition, game or pair row spans train and test; "
      "no outer-test game enters the inner CV")

fold assignments are identical to the LLM RQ1 pipeline's (955 rows, 5 repeats x 191 games)


CV assertions OK: no composition, game or pair row spans train and test; no outer-test game enters the inner CV


## 5. Held-out results: Strategy / Strategy Temporal / Combined

Each ordered pair receives one held-out probability per outer repeat. Those
five are **averaged per `game x voter x candidate` pair first**, and AUC/F1 are
computed once from the averages -- the five repeats are five looks at the same
2,882 pairs, not five independent observations. The per-repeat predictions are
kept only as a diagnostic.

Intervals come from a composition-clustered bootstrap: 1,000 resamples of the
36 player compositions with replacement, each carrying all of its games and
pair rows.

In [8]:
t0 = time.time()
results = {}
for name in hp.FEATURE_SETS:
    cols = hp.feature_set_columns(name)
    index, player_mat = hp.player_vectors(feat_df, cols)
    voter_ix, cand_ix = hp.pair_player_indices(pairs, index)
    oof, fold_meta = hm.run_nested_cv(pairs, player_mat, voter_ix, cand_ix, fold_df)
    avg = hm.average_oof(oof)
    results[name] = {"cols": cols, "player_mat": player_mat, "voter_ix": voter_ix,
                     "cand_ix": cand_ix, "oof": oof, "avg": avg,
                     "fold_meta": fold_meta,
                     "boot": hm.bootstrap_metrics(avg, n_boot=N_BOOT, seed=42)}
    print(f"{name:20s} fitted ({time.time() - t0:5.1f}s elapsed)")

human_fidelity_summary = pd.DataFrame([
    {"feature_set": name, "n_predictors": 2 * len(r["cols"]), **r["boot"]}
    for name, r in results.items()])
human_fidelity_summary.to_csv(TAB / "human_fidelity_summary.csv", index=False)
print()
print(human_fidelity_summary[["feature_set", "n_predictors", "n_pairs", "prevalence",
                              "auc", "auc_ci_lo", "auc_ci_hi",
                              "f1", "f1_ci_lo", "f1_ci_hi"]].round(4).to_string(index=False))

Strategy             fitted ( 11.9s elapsed)


Strategy Temporal    fitted ( 24.4s elapsed)


Combined             fitted ( 38.2s elapsed)

      feature_set  n_predictors  n_pairs  prevalence    auc  auc_ci_lo  auc_ci_hi     f1  f1_ci_lo  f1_ci_hi
         Strategy            12     2882       0.272 0.5330     0.5046     0.5641 0.3805    0.3472    0.4130
Strategy Temporal            24     2882       0.272 0.5420     0.5129     0.5698 0.3696    0.3382    0.4008
         Combined            26     2882       0.272 0.6439     0.5949     0.6889 0.4439    0.3881    0.4919


### Regularisation grid: was the optimum inside the searched range?

$C$ is inverse regularisation strength, so the grid spans $10^{-4}$ (very
heavy penalty) to $10^{4}$ (effectively unpenalised). If the inner CV
routinely selected an endpoint the grid would be censoring the answer, and
that has to be flagged rather than accepted.

In [9]:
penalty_rows = []
for name, r in results.items():
    report = hm.penalty_boundary_report(r["fold_meta"])
    print(f"{name:20s} boundary {report['pct_at_boundary']:5.1f}% "
          f"(lower {report['at_lower_bound']}, upper {report['at_upper_bound']} "
          f"of {report['n_fits']}), median C = {report['median_selected_C']:g}")
    fm = r["fold_meta"].copy()
    fm.insert(0, "feature_set", name)
    penalty_rows.append(fm)

selected_penalties = pd.concat(penalty_rows, ignore_index=True)
selected_penalties.to_csv(TAB / "selected_penalties.csv", index=False)

boundary_pct = max(hm.penalty_boundary_report(r["fold_meta"])["pct_at_boundary"]
                   for r in results.values())
if boundary_pct > 25:
    raise AssertionError(
        f"{boundary_pct:.0f}% of inner-CV fits selected a grid endpoint -- the C grid "
        "is censoring the optimum and must be widened before these results are used.")
print(f"\nworst boundary-selection rate {boundary_pct:.0f}% (< 25%): "
      "the grid is wide enough, the optimum is inside the searched range")

Strategy             boundary  12.0% (lower 1, upper 2 of 25), median C = 10
Strategy Temporal    boundary   4.0% (lower 1, upper 0 of 25), median C = 0.1
Combined             boundary   0.0% (lower 0, upper 0 of 25), median C = 0.01

worst boundary-selection rate 12% (< 25%): the grid is wide enough, the optimum is inside the searched range


### Paired feature-set comparisons

In [10]:
comparisons = pd.concat([
    hm.bootstrap_paired_difference(results["Strategy"]["avg"], results["Combined"]["avg"],
                                   "Strategy", "Combined", n_boot=N_BOOT, seed=42),
    hm.bootstrap_paired_difference(results["Strategy"]["avg"],
                                   results["Strategy Temporal"]["avg"],
                                   "Strategy", "Strategy Temporal",
                                   n_boot=N_BOOT, seed=42),
], ignore_index=True)
comparisons.to_csv(TAB / "human_feature_set_comparisons.csv", index=False)
print(comparisons.round(4).to_string(index=False))
print("\nBoth models are scored on the same resampled compositions, so these are "
      "paired differences. An interval that excludes zero is a bootstrap interval "
      "above/below zero, not a significance test.")

                  comparison metric  value_a  value_b  mean_diff   ci_lo  ci_hi  interval_excludes_zero  n_boot  n_boot_failed
         Combined - Strategy    auc   0.5330   0.6439     0.1109  0.0678 0.1508                    True    1000              0
         Combined - Strategy     f1   0.3805   0.4439     0.0635  0.0148 0.1055                    True    1000              0
Strategy Temporal - Strategy    auc   0.5330   0.5420     0.0090 -0.0137 0.0313                   False    1000              0
Strategy Temporal - Strategy     f1   0.3805   0.3696    -0.0109 -0.0386 0.0122                   False    1000              0

Both models are scored on the same resampled compositions, so these are paired differences. An interval that excludes zero is a bootstrap interval above/below zero, not a significance test.


## 6. Combined-model coefficients

The coefficient analysis is done once, for the aggregate **Combined** model
only. $C$ is selected by grouped CV on the complete dataset, the standardised
model is fitted on all pair rows, and the 26 coefficients are bootstrapped by
resampling whole player compositions 1,000 times, refitting the scaler and the
model inside each replicate with $C$ held fixed -- the same procedure as the
LLM RQ1 coefficient analysis.

The scaler is fitted on **unique player-game vectors**, not on the duplicated
pair rows, so a six-player game (30 pairs) cannot outweigh a four-player game
(12 pairs) in the standardisation, and the same feature is scaled identically
on the voter and candidate sides -- which is what makes the two blocks
directly comparable.

In [11]:
combined = results["Combined"]
cols = combined["cols"]
y = pairs["label"].to_numpy()
keys = pairs["key"].to_numpy(dtype=object)

final_C, final_cv_auc = hm.select_C_full_data(
    combined["player_mat"], combined["voter_ix"], combined["cand_ix"],
    y, keys, comp_of_key)
print(f"final C selected by grouped CV on the complete dataset: {final_C:g} "
      f"(mean CV AUC {final_cv_auc:.4f})")

final_model, X_final, _ = hm.fit_final_model(
    combined["player_mat"], combined["voter_ix"], combined["cand_ix"], y, final_C)

t0 = time.time()
boot_betas, boot_failures = hm.bootstrap_coefficients(
    pairs, combined["player_mat"], combined["voter_ix"], combined["cand_ix"],
    cols, final_C, n_boot=N_BOOT, seed=7)
print(f"coefficient bootstrap: {len(boot_betas)} of {N_BOOT} replicates in "
      f"{time.time() - t0:.1f}s, {len(boot_failures)} failures")
assert len(boot_failures) == 0, boot_failures

coef = (hm.coefficient_table(final_model, cols)
        .merge(hm.coefficient_bootstrap_table(boot_betas, cols, boot_failures),
               on=["side", "column"]))
coef["feature"] = [hp.rate_column_base(c) for c in coef["column"]]
coef["block"] = np.where(coef["feature"].isin(STRATEGY_FEATURES),
                         "strategy", "content and targets")
coef["selected_C"] = final_C
coef = coef[["side", "side_key", "block", "feature", "column", "beta", "odds_ratio",
             "beta_ci_lo", "beta_ci_hi", "odds_ratio_ci_lo", "odds_ratio_ci_hi",
             "selected_C", "n_boot_used", "n_boot_failed"]]
coef.to_csv(TAB / "human_combined_coefficients.csv", index=False)

boot_long = pd.DataFrame(
    boot_betas, columns=[f"{s}::{c}" for s in ("voter", "candidate") for c in cols])
boot_long.insert(0, "boot_iter", np.arange(len(boot_long)))
boot_long.to_csv(TAB / "human_combined_coefficient_bootstrap.csv", index=False)

print()
print(coef[["side", "feature", "beta", "odds_ratio",
            "beta_ci_lo", "beta_ci_hi"]].round(3).to_string(index=False))

final C selected by grouped CV on the complete dataset: 0.01 (mean CV AUC 0.6443)


coefficient bootstrap: 1000 of 1000 replicates in 11.6s, 0 failures

     side                        feature   beta  odds_ratio  beta_ci_lo  beta_ci_hi
    Voter                     accusation -0.035       0.965      -0.061      -0.011
    Voter                        defense  0.018       1.018      -0.039       0.066
    Voter                  interrogation -0.004       0.996      -0.047       0.030
    Voter                       evidence  0.021       1.021      -0.009       0.049
    Voter           identity_declaration -0.032       0.969      -0.065      -0.007
    Voter                call_for_action  0.016       1.016      -0.032       0.055
    Voter      werewolf_accusations_made -0.137       0.872      -0.182      -0.082
    Voter     deception_accusations_made -0.042       0.959      -0.083      -0.001
    Voter  werewolf_accusations_received  0.040       1.041       0.002       0.086
    Voter deception_accusations_received  0.005       1.005      -0.025       0.033
    Vot

The **candidate side** is the primary comparison with the LLM surrogate: an
LLM alternative corresponds most naturally to the player being considered for
a vote. Voter-side coefficients are reported for completeness but should be
read cautiously -- see the caveat in section 8.

Only direction, relative prominence and consistency of feature families are
compared with the LLM models. The human logistic coefficients and the LLM
conditional-logit coefficients are on different scales and are never compared
numerically.

In [12]:
llm = pd.read_csv(RQ1_TABLES / "ridge_coefficient_bootstrap_rates.csv")
llm_wide = llm.pivot(index="feature", columns="model", values="beta")
side_by_side = (coef[coef.side == "Candidate"]
                .set_index("feature")[["beta", "beta_ci_lo", "beta_ci_hi"]]
                .rename(columns={"beta": "human_beta", "beta_ci_lo": "human_lo",
                                 "beta_ci_hi": "human_hi"})
                .join(llm_wide.rename(columns=lambda m: f"llm_{m}_beta"))
                .loc[ALL_FEATURES_13])
side_by_side["same_sign_as_31B"] = (
    np.sign(side_by_side["human_beta"]) == np.sign(side_by_side["llm_31B_beta"]))
side_by_side.to_csv(TAB / "human_vs_llm_candidate_side.csv")
print("Candidate-side direction comparison (LLM betas are conditional-logit and "
      "are NOT numerically comparable -- sign and rank only):")
print(side_by_side.round(3).to_string())

Candidate-side direction comparison (LLM betas are conditional-logit and are NOT numerically comparable -- sign and rank only):
                                human_beta  human_lo  human_hi  llm_2B_beta  llm_31B_beta  llm_4B_beta  same_sign_as_31B
feature                                                                                                                 
accusation                          -0.066    -0.167     0.031        0.001        -0.003        0.018              True
defense                              0.034    -0.030     0.108        0.286         0.091        0.160              True
interrogation                        0.009    -0.084     0.092        0.049        -0.032       -0.028             False
evidence                            -0.103    -0.186    -0.004        0.005        -0.053       -0.089              True
identity_declaration                 0.008    -0.136     0.169       -0.110        -0.042       -0.137             False
call_for_action          

## 7. Figures

In [13]:
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

# Okabe-Ito, color-blind friendly
COLOR_STRATEGY = "#0072B2"          # blue
COLOR_CONTENT = "#D55E00"           # vermillion
COLOR_SET = {"Strategy": "#0072B2", "Strategy Temporal": "#E69F00",
             "Combined": "#009E73"}

# the human-readable names already used in the RQ1 thesis figures
FEATURE_LABELS = {
    "accusation": "Accusation", "defense": "Defense",
    "interrogation": "Interrogation", "evidence": "Evidence",
    "identity_declaration": "Identity Declaration",
    "call_for_action": "Call for Action",
    "werewolf_accusations_made": "Werewolf accusations made",
    "deception_accusations_made": "Deception accusations made",
    "werewolf_accusations_received": "Werewolf accusations received",
    "deception_accusations_received": "Deception accusations received",
    "claims_werewolf": "Claims Werewolf", "claims_tanner": "Claims Tanner",
    "claims_night_action_role": "Claims night-action role",
}
plt.rcParams.update({"font.size": 10, "axes.spines.top": False,
                     "axes.spines.right": False})

In [14]:
# --- Figure H1: Combined-model coefficients, voter and candidate panels -----
order = ALL_FEATURES_13
fig, axes = plt.subplots(1, 2, figsize=(11.5, 5.0), sharex=True)

for ax, side in zip(axes, ["Voter", "Candidate"]):
    sub = coef[coef.side == side].set_index("feature").loc[order]
    ypos = np.arange(len(order))[::-1]
    colors = [COLOR_STRATEGY if f in STRATEGY_FEATURES else COLOR_CONTENT
              for f in order]
    ax.errorbar(sub["beta"], ypos,
                xerr=[sub["beta"] - sub["beta_ci_lo"], sub["beta_ci_hi"] - sub["beta"]],
                fmt="none", ecolor="#666666", elinewidth=1.1, capsize=2.5, zorder=1)
    ax.scatter(sub["beta"], ypos, c=colors, s=42, zorder=2, edgecolor="white",
               linewidth=0.6)
    ax.axvline(0, color="black", lw=0.9, zorder=0)
    ax.set_yticks(ypos)
    ax.set_yticklabels([FEATURE_LABELS[f] for f in order])
    ax.set_title(f"{side} features", fontsize=12, pad=8)
    ax.set_xlabel("Standardised coefficient (log-odds)")
    ax.grid(axis="x", color="#e6e6e6", lw=0.7, zorder=0)

axes[1].tick_params(labelleft=False)
handles = [plt.Line2D([], [], marker="o", ls="", color=COLOR_STRATEGY, label="Strategy"),
           plt.Line2D([], [], marker="o", ls="", color=COLOR_CONTENT,
                      label="Content and targets")]
# bottom-right of the voter panel is empty; the shared x-axis is kept because
# it is the point of the figure that the voter block is small next to the
# candidate block
axes[0].legend(handles=handles, loc="lower right", frameon=False, fontsize=9)
fig.tight_layout()
fig.savefig(FIG / "human_combined_coefficients.png", dpi=300, bbox_inches="tight")
fig.savefig(FIG / "human_combined_coefficients.pdf", bbox_inches="tight")
plt.close(fig)
print("saved", (FIG / "human_combined_coefficients.png").relative_to(REPO_ROOT))

saved analysis\human_outcomes\pairwise_logreg_v2\figures\human_combined_coefficients.png


In [15]:
# --- Figure H2: held-out AUC by feature set --------------------------------
fig, ax = plt.subplots(figsize=(6.2, 3.6))
names = list(hp.FEATURE_SETS)
s = human_fidelity_summary.set_index("feature_set").loc[names]
x = np.arange(len(names))
ax.bar(x, s["auc"], 0.55, color=[COLOR_SET[n] for n in names],
       yerr=[s["auc"] - s["auc_ci_lo"], s["auc_ci_hi"] - s["auc"]],
       capsize=4, ecolor="#444444")
ax.axhline(0.5, color="black", lw=0.9, ls="--")
ax.text(len(names) - 0.45, 0.503, "no information", fontsize=8, ha="right", va="bottom")
ax.set_xticks(x)
ax.set_xticklabels(names)
ax.set_ylabel("Held-out ROC-AUC")
ax.set_ylim(0.45, max(0.75, s["auc_ci_hi"].max() + 0.03))
ax.grid(axis="y", color="#e6e6e6", lw=0.7)
ax.set_axisbelow(True)
fig.tight_layout()
fig.savefig(FIG / "human_feature_set_auc.png", dpi=300, bbox_inches="tight")
fig.savefig(FIG / "human_feature_set_auc.pdf", bbox_inches="tight")
plt.close(fig)
print("saved", (FIG / "human_feature_set_auc.png").relative_to(REPO_ROOT))

saved analysis\human_outcomes\pairwise_logreg_v2\figures\human_feature_set_auc.png


## 8. Save outputs

In [16]:
# tables/
lai_replication.to_csv(TAB / "lai_replication.csv", index=False)
lai_counts.to_csv(TAB / "lai_split_row_counts.csv", index=False)
fold_df.drop(columns=["key"]).to_csv(TAB / "group_cv_assignments.csv", index=False)
support_check.to_csv(TAB / "human_feature_support.csv", index=False)

oof_out = []
for name, r in results.items():
    a = r["avg"].copy()
    a.insert(0, "feature_set", name)
    oof_out.append(a)
human_oof_predictions = pd.concat(oof_out, ignore_index=True)
human_oof_predictions.to_csv(TAB / "human_oof_predictions.csv", index=False)

# diagnostics/
diagnostics["unresolved_names"].to_csv(DIA / "unresolved_names.csv", index=False)
unresolved.to_csv(DIA / "unresolved_name_summary.csv", index=False)
zero_denominator.to_csv(DIA / "zero_denominator_report.csv", index=False)
conservation.to_csv(DIA / "conservation_report.csv", index=False)
excluded_voters.to_csv(DIA / "excluded_voters.csv", index=False)
pd.concat([r["oof"].assign(feature_set=n) for n, r in results.items()],
          ignore_index=True).to_csv(DIA / "human_oof_predictions_per_repeat.csv",
                                    index=False)

print("tables:")
for p in sorted(TAB.glob("*.csv")):
    print("  ", p.name)
print("figures:")
for p in sorted(FIG.iterdir()):
    print("  ", p.name)
print("diagnostics:")
for p in sorted(DIA.glob("*.csv")):
    print("  ", p.name)

tables:
   group_cv_assignments.csv
   human_combined_coefficient_bootstrap.csv
   human_combined_coefficients.csv
   human_feature_set_comparisons.csv
   human_feature_support.csv
   human_fidelity_summary.csv
   human_oof_predictions.csv
   human_vs_llm_candidate_side.csv
   lai_replication.csv
   lai_split_row_counts.csv
   selected_penalties.csv
figures:
   human_combined_coefficients.pdf
   human_combined_coefficients.png
   human_feature_set_auc.pdf
   human_feature_set_auc.png
diagnostics:
   conservation_report.csv
   excluded_voters.csv
   human_oof_predictions_per_repeat.csv
   integrity_checks.csv
   unresolved_name_summary.csv
   unresolved_names.csv
   zero_denominator_report.csv


## 9. Integrity checks

Everything asserted above is re-stated here as one pass/fail block, plus the
numerical and determinism checks.

In [17]:
checks = []


def check(name, ok, detail=""):
    checks.append({"check": name, "passed": bool(ok), "detail": detail})
    assert ok, f"FAILED: {name} -- {detail}"


c = hp.corpus_counts(roster)
check("corpus: 191 cleaned ONUW games", c["games"] == 191, str(c["games"]))
check("corpus: 60 unique players", c["unique_players"] == 60, str(c["unique_players"]))
check("corpus: 36 unique compositions", c["unique_compositions"] == 36,
      str(c["unique_compositions"]))

per_voter = pairs.groupby(["key", "voter"]).agg(n=("label", "size"),
                                                pos=("label", "sum"),
                                                np_=("n_players", "first"))
check("pairwise: exactly n-1 candidate rows per voter", (per_voter["n"] == per_voter["np_"] - 1).all())
check("pairwise: exactly one positive per voter", (per_voter["pos"] == 1).all())
check("pairwise: voter != candidate", (pairs["voter"] != pairs["candidate"]).all())
check("pairwise: chosen candidate is on the roster",
      all(set(g["candidate"]) <= set(roster[k]["players"]) for k, g in pairs.groupby("key")))
check("pairwise: 2882 rows / 784 positives",
      (len(pairs), int(pairs["label"].sum())) == (2882, 784),
      f"{len(pairs)} rows, {int(pairs['label'].sum())} positives")

check("features: totals identical to the RQ1 feature table", support_check["matches_rq1"].all())
check("features: early + late == overall", conservation["conserved"].all())
check("features: unresolved names identical to RQ1",
      set(unresolved["raw_name"]) == {"Jordan", "Kayla", "Matt"})
check("features: no ground-truth role or outcome in the feature list", not _leaks)

check("cv: fold assignments identical to the LLM RQ1 pipeline", True)
check("cv: no composition, game or pair spans train and test", True)
check("cv: no outer-test game enters the inner CV", True)

for name, r in results.items():
    p = r["avg"]["prob"].to_numpy()
    check(f"numerical: {name} held-out probabilities finite and in [0, 1]",
          np.isfinite(p).all() and ((p >= 0) & (p <= 1)).all())
    check(f"numerical: {name} every logistic fit converged", True,
          "asserted inside hm.fit_logreg")
check("numerical: 0 bootstrap failures in the fidelity intervals",
      all(r["boot"]["n_boot_failed"] == 0 for r in results.values()))
check("numerical: 0 bootstrap failures in the coefficient bootstrap",
      len(boot_failures) == 0)
check("numerical: selected C never censored at a grid endpoint (< 25%)",
      boundary_pct < 25, f"{boundary_pct:.0f}%")

# determinism: refit the cheapest feature set and require identical predictions
_cols = hp.feature_set_columns("Strategy")
_idx, _mat = hp.player_vectors(feat_df, _cols)
_v, _c = hp.pair_player_indices(pairs, _idx)
_oof2, _fm2 = hm.run_nested_cv(pairs, _mat, _v, _c, fold_df)
_a2 = hm.average_oof(_oof2)
check("determinism: a second nested-CV run reproduces every held-out probability",
      np.allclose(_a2["prob"].to_numpy(), results["Strategy"]["avg"]["prob"].to_numpy(),
                  rtol=0, atol=0))
check("determinism: a second run selects the same C in every fold",
      (_fm2["selected_C"].to_numpy() == results["Strategy"]["fold_meta"]["selected_C"].to_numpy()).all())

integrity = pd.DataFrame(checks)
integrity.to_csv(DIA / "integrity_checks.csv", index=False)
print(integrity.to_string(index=False))
print(f"\nall {len(integrity)} integrity checks passed")

                                                                    check  passed                        detail
                                           corpus: 191 cleaned ONUW games    True                           191
                                                corpus: 60 unique players    True                            60
                                           corpus: 36 unique compositions    True                            36
                           pairwise: exactly n-1 candidate rows per voter    True                              
                                 pairwise: exactly one positive per voter    True                              
                                             pairwise: voter != candidate    True                              
                              pairwise: chosen candidate is on the roster    True                              
                                      pairwise: 2882 rows / 784 positives    True      2882 rows, 784 po

### Caveats for the thesis

1. **Composition grouping is not participant independence.** 22 of the 60
   players appear in more than one composition, so an individual can still
   fall on both sides of a split. Inherent to the corpus; not solvable here.
2. **Voter-side coefficients are weakly identified.** Every voter contributes
   exactly one positive among their $n-1$ rows, so voter features cannot
   discriminate *which* candidate is chosen -- they can only track the
   base rate, which varies with roster size ($1/3$ in a four-player game,
   $1/5$ in a six-player one). Voter-side coefficients therefore partly absorb
   roster-size variation and should not be read as preferences. The
   candidate-side block is the interpretable one, and is also the one that
   corresponds to an LLM alternative.
3. **Seven games contribute no voter rows**, because every voter in them
   abstained, moderated, or voted for themselves. They remain in the corpus
   and the fold map, but the pairwise dataset covers 184 games.
4. **The replication is a replication, not an exact reproduction.** The row
   counts match to the row and the hyperparameters come from the authors'
   released code, but their published AUC uses hard labels, and their exact
   random seed and sklearn version are unknown.
5. **Associations, not effects.** Regularised coefficients from an
   observational corpus; bootstrap intervals are intervals, not tests.